File: Sources/DayPlan/Integrations/Todoist/TodoistAPIClient.swift
import Foundation

final class TodoistAPIClient: TodoistTaskProviding, @unchecked Sendable {
    private let credentialStore: any CredentialStore
    private let session: URLSession
    private let baseURL = URL(string: "https://api.todoist.com/api/v1")

    init(credentialStore: any CredentialStore, session: URLSession = .shared) {
        self.credentialStore = credentialStore
        self.session = session
    }

    func listTasks(limit: Int = 100, projectID: String? = nil) async throws -> [TodoistTask] {
        let token = try await accessToken()
        var tasks: [TodoistTask] = []
        var cursor: String?
        let boundedLimit = min(max(limit, 1), 100)

        repeat {
            var components = URLComponents(
                url: try endpoint("tasks"), resolvingAgainstBaseURL: false)
            var queryItems = [
                URLQueryItem(name: "limit", value: String(min(100, boundedLimit - tasks.count)))
            ]
            if let projectID, !projectID.isEmpty {
                queryItems.append(URLQueryItem(name: "project_id", value: projectID))
            }
            if let cursor {
                queryItems.append(URLQueryItem(name: "cursor", value: cursor))
            }
            components?.queryItems = queryItems
            guard let url = components?.url else { throw TodoistAPIError.invalidResponse }

            let data = try await sendRequest(url: url, method: "GET", token: token)
            let page = try JSONDecoder().decode(TodoistTaskPage.self, from: data)
            tasks.append(contentsOf: page.results)
            cursor = page.nextCursor
        } while cursor != nil && tasks.count < boundedLimit

        return tasks
    }

    func getTask(id: String) async throws -> TodoistTask {
        let token = try await accessToken()
        let data = try await sendRequest(url: endpoint("tasks/\(id)"), method: "GET", token: token)
        do { return try JSONDecoder().decode(TodoistTask.self, from: data) } catch {
            throw TodoistAPIError.invalidResponse
        }
    }

    func listProjects() async throws -> [TodoistProject] {
        let token = try await accessToken()
        var projects: [TodoistProject] = []
        var cursor: String?

        repeat {
            var components = URLComponents(
                url: try endpoint("projects"), resolvingAgainstBaseURL: false)
            var queryItems = [URLQueryItem(name: "limit", value: "100")]
            if let cursor {
                queryItems.append(URLQueryItem(name: "cursor", value: cursor))
            }
            components?.queryItems = queryItems
            guard let url = components?.url else { throw TodoistAPIError.invalidResponse }

            let data = try await sendRequest(url: url, method: "GET", token: token)
            let page = try JSONDecoder().decode(TodoistProjectPage.self, from: data)
            projects.append(contentsOf: page.results)
            cursor = page.nextCursor
        } while cursor != nil

        return projects.sorted {
            if $0.isInbox != $1.isInbox { return $0.isInbox }
            return $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
        }
    }

    func listLabels() async throws -> [TodoistLabel] {
        let token = try await accessToken()
        var labels: [TodoistLabel] = []
        var cursor: String?

        repeat {
            var components = URLComponents(
                url: try endpoint("labels"), resolvingAgainstBaseURL: false)
            var queryItems = [URLQueryItem(name: "limit", value: "100")]
            if let cursor {
                queryItems.append(URLQueryItem(name: "cursor", value: cursor))
            }
            components?.queryItems = queryItems
            guard let url = components?.url else { throw TodoistAPIError.invalidResponse }

            let data = try await sendRequest(url: url, method: "GET", token: token)
            let page = try JSONDecoder().decode(TodoistLabelPage.self, from: data)
            labels.append(contentsOf: page.results)
            cursor = page.nextCursor
        } while cursor != nil

        return labels.sorted {
            $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
        }
    }

    func createTask(draft: TodoistTaskDraft) async throws -> TodoistTask {
        let token = try await accessToken()
        let url = try endpoint("tasks")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 20
        request.httpBody = try JSONEncoder().encode(
            TodoistCreateTaskRequest(
                content: draft.content,
                description: draft.description.isEmpty ? nil : draft.description,
                projectID: draft.projectID,
                dueDate: draft.dueDate,
                priority: draft.priority,
                labels: draft.labels
            )
        )

        let data = try await send(request)
        do {
            return try JSONDecoder().decode(TodoistTask.self, from: data)
        } catch {
            throw TodoistAPIError.invalidResponse
        }
    }

    func completeTask(id: String) async throws {
        let token = try await accessToken()
        let url = try endpoint("tasks/\(id)/close")
        _ = try await sendRequest(url: url, method: "POST", token: token)
    }

    func reopenTask(id: String) async throws {
        let token = try await accessToken()
        let url = try endpoint("tasks/\(id)/reopen")
        _ = try await sendRequest(url: url, method: "POST", token: token)
    }

    func deleteTask(id: String) async throws {
        let token = try await accessToken()
        let url = try endpoint("tasks/\(id)")
        _ = try await sendRequest(url: url, method: "DELETE", token: token)
    }

    func updateTask(id: String, patch: TodoistTaskPatch) async throws -> TodoistTask {
        let token = try await accessToken()
        let url = try endpoint("tasks/\(id)")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 20
        request.httpBody = try JSONEncoder().encode(TodoistUpdateTaskRequest(patch: patch))
        let data = try await send(request)
        do { return try JSONDecoder().decode(TodoistTask.self, from: data) } catch {
            throw TodoistAPIError.invalidResponse
        }
    }

    private func accessToken() async throws -> String {
        guard let token = try await credentialStore.readCredential(for: .todoist), !token.isEmpty
        else {
            throw TodoistAPIError.credentialMissing
        }
        return token
    }

    private func endpoint(_ path: String) throws -> URL {
        guard let baseURL else { throw TodoistAPIError.invalidResponse }
        return baseURL.appending(path: path)
    }

    private func sendRequest(url: URL, method: String, token: String) async throws -> Data {
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 20
        return try await send(request)
    }

    private func send(_ request: URLRequest) async throws -> Data {
        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw TodoistAPIError.networkUnavailable
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw TodoistAPIError.invalidResponse
        }
        switch httpResponse.statusCode {
        case 200..<300:
            return data
        case 401:
            throw TodoistAPIError.unauthorized
        default:
            throw TodoistAPIError.serverResponse(httpResponse.statusCode)
        }
    }
}

struct TodoistUpdateTaskRequest: Encodable {
    let patch: TodoistTaskPatch

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try encode(patch.content, key: .content, in: &container)
        try encode(patch.description, key: .description, in: &container)
        try encode(patch.dueDate, key: .dueDate, in: &container)
        try encode(patch.priority, key: .priority, in: &container)
        try encode(patch.labels, key: .labels, in: &container)
    }

    private func encode<Value: Encodable>(
        _ field: TaskField<Value>, key: CodingKeys,
        in container: inout KeyedEncodingContainer<CodingKeys>
    ) throws {
        switch field {
        case .unchanged: break
        case .clear: try container.encodeNil(forKey: key)
        case .set(let value): try container.encode(value, forKey: key)
        }
    }

    private enum CodingKeys: String, CodingKey {
        case content
        case description
        case dueDate = "due_date"
        case priority
        case labels
    }
}

private struct TodoistTaskPage: Decodable {
    let results: [TodoistTask]
    let nextCursor: String?

    private enum CodingKeys: String, CodingKey {
        case results
        case nextCursor = "next_cursor"
    }
}

private struct TodoistProjectPage: Decodable {
    let results: [TodoistProject]
    let nextCursor: String?

    private enum CodingKeys: String, CodingKey {
        case results
        case nextCursor = "next_cursor"
    }
}

private struct TodoistLabelPage: Decodable {
    let results: [TodoistLabel]
    let nextCursor: String?

    private enum CodingKeys: String, CodingKey {
        case results
        case nextCursor = "next_cursor"
    }
}

struct TodoistCreateTaskRequest: Encodable {
    let content: String
    let description: String?
    let projectID: String?
    let dueDate: String?
    let priority: Int
    let labels: [String]

    private enum CodingKeys: String, CodingKey {
        case content
        case description
        case projectID = "project_id"
        case dueDate = "due_date"
        case priority
        case labels
    }
}

File: src/main/integrations/todoist/TodoistApiClient.ts
import type { CredentialService } from '../../security/CredentialService'
import type { TaskDraft, TaskPatch, TodoistLabel, TodoistProject, TodoistTask } from '../../../shared/domain'

interface Page<T> {
  results: T[]
  next_cursor: string | null
}

interface CompletedTasksPage<T> {
  items: T[]
  next_cursor: string | null
}

export class TodoistApiError extends Error {
  constructor(message: string, readonly statusCode?: number) {
    super(message)
    this.name = 'TodoistApiError'
  }
}

export class TodoistApiClient {
  private readonly baseUrl = 'https://api.todoist.com/api/v1'

  constructor(
    private readonly credentialService: CredentialService,
    private readonly request: typeof fetch = fetch,
  ) {}

  async testConnection(workspaceId: string, candidateToken?: string): Promise<void> {
    await this.listTasks(workspaceId, { limit: 1 }, candidateToken)
  }

  async listTasks(workspaceId: string, options: { limit?: number; project_id?: string } = {}, candidateToken?: string): Promise<TodoistTask[]> {
    const limit = Math.min(Math.max(options.limit ?? 200, 1), 5000)
    const tasks: TodoistTask[] = []
    let cursor: string | null = null
    const seenCursors = new Set<string>()

    do {
      const url = new URL(`${this.baseUrl}/tasks`)
      url.searchParams.set('limit', String(Math.min(200, limit - tasks.length)))
      if (options.project_id) url.searchParams.set('project_id', options.project_id)
      if (cursor) url.searchParams.set('cursor', cursor)
      const page = await this.get<Page<TodoistTask>>(url, workspaceId, candidateToken)
      tasks.push(...page.results)
      cursor = page.next_cursor
      if (cursor && seenCursors.has(cursor)) throw new TodoistApiError('Todoist returned invalid pagination data.')
      if (cursor) seenCursors.add(cursor)
    } while (cursor && tasks.length < limit)

    return tasks.slice(0, limit)
  }

  async getTask(workspaceId: string, taskId: string): Promise<TodoistTask> {
    return this.get<TodoistTask>(`${this.baseUrl}/tasks/${encodeURIComponent(taskId)}`, workspaceId)
  }

  async listProjects(workspaceId: string): Promise<TodoistProject[]> {
    return this.getAllPages<TodoistProject>('/projects', workspaceId)
  }

  async listLabels(workspaceId: string): Promise<TodoistLabel[]> {
    return this.getAllPages<TodoistLabel>('/labels', workspaceId)
  }

  async createTask(workspaceId: string, draft: TaskDraft): Promise<TodoistTask> {
    const body: Record<string, unknown> = {
      content: draft.content.trim(),
      description: draft.description?.trim() ?? '',
      priority: draft.priority ?? 1,
      labels: draft.labels ?? [],
    }
    if (draft.project_id) body.project_id = draft.project_id
    if (draft.due_date) body.due_date = draft.due_date
    return this.send<TodoistTask>(`${this.baseUrl}/tasks`, 'POST', workspaceId, body)
  }

  async updateTask(workspaceId: string, taskId: string, patch: TaskPatch): Promise<TodoistTask> {
    const body: Record<string, unknown> = {}
    if (patch.content !== undefined) body.content = patch.content.trim()
    if (patch.description !== undefined) body.description = patch.description.trim()
    if (patch.due_date !== undefined) body.due_date = patch.due_date
    if (patch.priority !== undefined) body.priority = patch.priority
    if (patch.labels !== undefined) body.labels = patch.labels
    return this.send<TodoistTask>(`${this.baseUrl}/tasks/${encodeURIComponent(taskId)}`, 'POST', workspaceId, body)
  }

  async completeTask(workspaceId: string, taskId: string): Promise<void> {
    await this.send<void>(`${this.baseUrl}/tasks/${encodeURIComponent(taskId)}/close`, 'POST', workspaceId)
  }

  async reopenTask(workspaceId: string, taskId: string): Promise<void> {
    await this.send<void>(`${this.baseUrl}/tasks/${encodeURIComponent(taskId)}/reopen`, 'POST', workspaceId)
  }

  async deleteTask(workspaceId: string, taskId: string): Promise<void> {
    await this.send<void>(`${this.baseUrl}/tasks/${encodeURIComponent(taskId)}`, 'DELETE', workspaceId)
  }

  async listCompletedTasks(workspaceId: string, since: string, until: string): Promise<TodoistTask[]> {
    const tasks: TodoistTask[] = []
    let cursor: string | null = null
    const seenCursors = new Set<string>()
    do {
      const url = new URL(`${this.baseUrl}/tasks/completed/by_completion_date`)
      url.searchParams.set('since', since)
      url.searchParams.set('until', until)
      url.searchParams.set('limit', '200')
      if (cursor) url.searchParams.set('cursor', cursor)
      const page = await this.get<CompletedTasksPage<TodoistTask>>(url, workspaceId)
      tasks.push(...page.items)
      cursor = page.next_cursor
      if (cursor && seenCursors.has(cursor)) throw new TodoistApiError('Todoist returned invalid pagination data.')
      if (cursor) seenCursors.add(cursor)
    } while (cursor)
    return tasks
  }

  private async getAllPages<T>(path: string, workspaceId: string): Promise<T[]> {
    const items: T[] = []
    let cursor: string | null = null
    const seenCursors = new Set<string>()
    do {
      const url = new URL(`${this.baseUrl}${path}`)
      url.searchParams.set('limit', '200')
      if (cursor) url.searchParams.set('cursor', cursor)
      const page = await this.get<Page<T>>(url, workspaceId)
      items.push(...page.results)
      cursor = page.next_cursor
      if (cursor && seenCursors.has(cursor)) throw new TodoistApiError('Todoist returned invalid pagination data.')
      if (cursor) seenCursors.add(cursor)
    } while (cursor)
    return items
  }

  private async get<T>(url: URL | string, workspaceId: string, candidateToken?: string): Promise<T> {
    return this.send<T>(url, 'GET', workspaceId, undefined, candidateToken)
  }

  private async send<T>(url: URL | string, method: string, workspaceId: string, body?: unknown, candidateToken?: string): Promise<T> {
    const token = candidateToken ?? await this.credentialService.readTodoistToken(workspaceId)
    if (!token) throw new TodoistApiError('Add your Todoist API token in Settings.')

    let response: Response
    try {
      response = await this.request(url, {
        method,
        headers: {
          Authorization: `Bearer ${token}`,
          ...(body === undefined ? {} : { 'Content-Type': 'application/json' }),
        },
        ...(body === undefined ? {} : { body: JSON.stringify(body) }),
        signal: AbortSignal.timeout(20000),
      })
    } catch {
      throw new TodoistApiError('Could not reach Todoist. Check your internet connection.')
    }

    if (response.status === 401) throw new TodoistApiError('Todoist rejected this token. Check it in Settings.', 401)
    if (response.status === 404) throw new TodoistApiError('The Todoist item could not be found.', 404)
    if (!response.ok) throw new TodoistApiError(`Todoist returned an error (${response.status}).`, response.status)
    if (response.status === 204 || response.headers.get('content-length') === '0') return undefined as T

    try {
      return await response.json() as T
    } catch {
      throw new TodoistApiError('Todoist returned an unreadable response.')
    }
  }
}
